File: 백준/Gold/5557. 1학년/1학년.py
from sys import *
from collections import * 
from heapq import * 
from functools import *
from math import *
from itertools import *

#setrecursionlimit(10**9)
input = stdin.readline


n = int(input())
arr = [0] +  list(map(int,input().split()))

dp = [[0 for _ in range(21)] for _ in range(n+1) ] #dp[i][j] = i번째까지 연산으로 j를 만들 수 있는 가능성


# 음수 및 20 초과 안됨 

dp[1][arr[1]] = 1 # 1번째 값은 뻬면 음수이므로 , 덧셈만 가능 
for i in range(2,n+1): # 연산한 구간 
    for j in range(21): # 20까지

        if dp[i-1][j] == 0 : # 이전 연산 개수가 0이면 불가능
            continue
        
        nxt1 = j + arr[i] # 다음 합 연산 결과
        nxt2 = j - arr[i] # 다음 차 연산 결과
        
        if nxt1 <= 20: # 20보다 작고
            dp[i][nxt1] += dp[i-1][j]
        
        if nxt2 >= 0: # 음수가 아닐 때
            dp[i][nxt2] += dp[i-1][j]


print(dp[n-1][arr[n]]) # 마지막 수를 제외한 모든 연산을 끝 맞췄을 때 마지막 숫자가 와 같은 총 개수 출력

File: 백준/Gold/5557. 1학년/1학년.swift
let n = Int(readLine()!)!

var numbers = readLine()!.split{$0 == " "}.map{Int($0)!}

var dp = [[Int]](repeating: [Int](repeating: 0, count: 21), count: n)

dp[0][numbers[0]] = 1

for i in 1..<n {
    
    var now = numbers[i]
    
    for j in 0...20 {
        
        if now <= j {
            
            if dp[i-1][j-now] != 0 {
                dp[i][j] += dp[i-1][j-now] // now를 더할 때
            }
        }
        
        if now+j <= 20 {
            dp[i][j] += dp[i-1][j+now] // now를 뺄 때
        }
        
        
        
        
    }
    
}

print(dp[n-2][numbers[n-1]]) // n-2 까지 써서 n-1을 만들 수 있는 경우의 수
